[5.1] Fix for the Pagination Query that selected all rows from a table and would fail on large datasets

mxaddict opened the pull request:
Implemented my 4.2.x Fix for the pagination count query into 5.1.x

Fix for the Pagination Query that selected all rows from a table and would fail on large datasets

mxaddict committed:
Updated pagination count logic to use count query instead of PHP count
function after selecting all rows from database

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,33 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
-
-        $this->aggregate = null;
+        if ($grouped)
+        {
```
Please follow PSR2.

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,33 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
     /**
      * Backup some fields for the pagination count.
      *
+     * @param bool $grouped
```
Please put **2** spaces either side of `bool`.

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1461,11 +1477,14 @@ protected function backupFieldsForCount()
     /**
      * Restore some fields after the pagination count.
      *
+     * @param bool $grouped
```
same

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -169,6 +169,20 @@ class Builder
+    protected $backupFields = ['orders', 'limit', 'offset', 'columns'];
+
+    /**
+     * These are additional fields we backup when doing counts with groups
```
Please end the description in a fullstop.

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -169,6 +169,20 @@ class Builder
     protected $backups = [];
 
     /**
+     * These are the fields that we backup when doing counts
```
Please end the description in a fullstop.

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,33 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
     public function getCountForPagination($columns = ['*'])
     {
-        $this->backupFieldsForCount();
+        $grouped = isset($this->groups) && $selectSql = $this->toSql();
```
Please do the second variable assignment first, above.

mxaddict commented:
Updated the commit to have the changes you requested :smile:

mxaddict commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,33 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
-
-        $this->aggregate = null;
+        if ($grouped) {
+            $this->from = $this->raw("({$selectSql}) as subquery");
```
BTW, Is this line OK? I was concerned that It might need changing...

GrahamCampbell replied:
Looks ok.

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -169,6 +169,20 @@ class Builder
     /**
+     * These are the fields that we backup when doing counts.
+     *
+     * @var string
```
not string, array

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -169,6 +169,20 @@ class Builder
+    /**
+     * These are additional fields we backup when doing counts with groups.
+     *
+     * @var string
```
same

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,33 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
     public function getCountForPagination($columns = ['*'])
     {
-        $this->backupFieldsForCount();
+        $selectSql = $this->toSql();
```
This line can surely be moved inside the if block so it's not executed if we don't need it?

mxaddict replied:
It will stop working if it's inside the if block, cause it needs to be the query before backing up the fields.

mxaddict replied:
Wait, I think I have a solution, I will update.

GrahamCampbell replied:
Oh right, I see. We're going to need two if blocks then. :)

mxaddict replied:
I think the original line:

```php
$grouped = isset($this->groups) && $selectSql = $this->toSql();
```

Did what you wanted, cause it will not Assign $selectSql if isset($this->groups) is false :smile:

GrahamCampbell replied:
Yeh, I know that does that. I want to keep that behaviour, but not use that style.

GrahamCampbell replied:
```php
if ($grouped = isset($this->groups)) {
    $selectSql = $this->toSql();
}
```
would do the job

GrahamCampbell commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -169,6 +169,20 @@ class Builder
+     *
+     * @var array
+     */
+    protected $backupFieldsMore = ['from', 'wheres', 'groups'];
```
backupMoreFields would be better

mxaddict commented:
Added the changes

GrahamCampbell commented:
Ping @codelight89, @fcjailybo, @JoostK, @codelight89, and @jarektkaczyk for review. This function was last modified in #8310 which came out of #8113, #8112 and #7372 if we go all the way back. :)

GrahamCampbell commented:
@mxaddict could you use a better commit message please. Don't mention versions, just describe what it does. We then need to wait for some reviews. It's important we're sure this is good. :)

mxaddict commented:
@GrahamCampbell updated the commit message :smile:

RomainLanz commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1427,31 +1441,34 @@ public function simplePaginate($perPage = 15, $columns = ['*'], $pageName = 'pag
+        }
 
-        $this->restoreFieldsForCount();
+        $total = $this->count($columns);
```
Do we realy need to instantiate a variable here?

mxaddict replied:
I would think so, cause we can't return the value before we restore the backed up fields. Unless you have a way of restoring the fields and returning the value on a single line, I think this will be necessary

taylorotwell commented:
Please send again when it has unit tests.

mxaddict commented:
@taylorotwell, this was an existing feature and already has tests here:

tests/Database/DatabaseEloquentIntegrationTest.php in the testCountForPaginationWithGrouping Method

mxaddict commented:
So what kind of test do I need to add? Cause the functionality is already tested in the original integration test. Let me know so that I can create one :smile:

GrahamCampbell commented:
Taylor, this already has unit tests. They were put it earlier when my PR was merged.

GrahamCampbell commented:
This is just a change in the implementation for performance reasons rather than a bug fix.

mxaddict commented:
I Added more fields to the backupMoreFields array, cause they were not needing for the agregate and were causing slowness with union and join queries that used pagination.

@GrahamCampbell what do you think about the new list of backup fields?

mxaddict commented:
@GrahamCampbell , I looked more into the current integration test, and I think the coverage is lacking, would it be OK for me to add a couple of tests to better cover the use-cases for the pagination count query?

Cause I might as well add it now while I'm fresh from making the changes.

mxaddict commented:
@GrahamCampbell, I implemented a couple of tests to make sure that the counts work as expected when using normal flow, count without grouping, count with grouping, count with grouping + joins, count with grouping + leftJoins, I tried to follow the way other tests were implemented, but let me know if I need to change/update anything.

Also, I'm not sure if the file I added the tests to is correct location, so let me know if I need to move them elsewhere.

beanmoss commented:
:+1:

zot24 commented:
:+1:

overtrue commented:
:+1:

GrahamCampbell commented:
Have you guys actually tested this out?

mxaddict commented:
@GrahamCampbell, I would advise we not merge this just now, I'm not sure that it's 100% A OK, can we ask for more input from other contributors?

GrahamCampbell commented:
Ping @codelight89, @fcjailybo, @JoostK, @codelight89, and @jarektkaczyk for review.

taylorotwell commented:
Let's re-open this when we are more confident then.

giacomoburattini commented:
We had the same issue and we fixed it in a similar way,  this is a big deal with large datasets. Watching this thread since we are interested in knowing when this pull request will be merged.

mxaddict commented:
I don't think this will be merged anytime soon, We know the issue and know the solution, but I myself am not familiar with the process of making and getting a PR merged into this project, So i have left it for now, been busy with other stuff, sorry @giacomoburattini
